### PR TITLE
added unit tests for spending module

### PR DESCRIPTION
**Pull request by dkijania**
incorporated new way of dealing with spending counter
added scenario  tests

### CODE REVIEW

File: chain-impl-mockchain/src/accounting/account/spending.rs [hunk 1]
```diff
@@ -153,3 +153,114 @@ impl From<SpendingCounter> for u32 {
...
+        assert_eq!(counter + 1, spending_counter.unlaned_counter());
+    }
+
+    #[quickcheck_macros::quickcheck]
+    pub fn increment_nth(mut spending_counter: SpendingCounter, n: u32) {
+        let lane_before = spending_counter.lane();
+        let counter = spending_counter.unlaned_counter();
+        spending_counter = spending_counter.increment_nth(n);
+        assert_eq!(lane_before, spending_counter.lane());
+        assert_eq!(counter + n, spending_counter.unlaned_counter());
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/accounting/account/spending.rs:201]: same as above, but we should also account for the possibility that `counter + n` does not fit in a u32 and panics in debug mode

File: chain-impl-mockchain/src/accounting/account/spending.rs [hunk 2]
```diff
@@ -153,3 +153,114 @@ impl From<SpendingCounter> for u32 {
...
+        assert_eq!(counter, sc.unlaned_counter());
+    }
+
+    #[quickcheck_macros::quickcheck]
+    fn increment_counter(mut spending_counter: SpendingCounter) {
+        let lane_before = spending_counter.lane();
+        let counter = spending_counter.unlaned_counter();
+        spending_counter = spending_counter.increment();
+        assert_eq!(lane_before, spending_counter.lane());
+        assert_eq!(counter + 1, spending_counter.unlaned_counter());
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/accounting/account/spending.rs:192]: this should be a wrapping add, as that's how the spending counter deals with overflows in the counter

File: chain-impl-mockchain/src/accounting/account/spending.rs [hunk 3]
```diff
@@ -153,3 +153,114 @@ impl From<SpendingCounter> for u32 {
...
+    }
+
+    #[test]
+    pub fn spending_counters_incorrect_order() {
+        let counters = vec![SpendingCounter::new(1, 0), SpendingCounter::new(0, 0)];
+        assert!(SpendingCounterIncreasing::new_from_counters(counters).is_none());
+    }
+
+    #[test]
+    pub fn spending_counters_overflow() {
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/accounting/account/spending.rs:227]: nit: I don't get the naming here

File: chain-impl-mockchain/src/accounting/account/spending.rs [hunk 4]
```diff
@@ -153,3 +153,114 @@ impl From<SpendingCounter> for u32 {
...
+
+    #[test]
+    #[should_panic]
+    pub fn spending_counter_increasing_wrong_lane() {
+        let mut sc_increasing = correct_spending_counter_increasing();
+        let incorrect_sc = SpendingCounter::new(SpendingCounterIncreasing::LANES, 1);
+        assert!(sc_increasing.next_verify(incorrect_sc).is_err());
+    }
+
+    fn correct_spending_counter_increasing() -> SpendingCounterIncreasing {
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/accounting/account/spending.rs:260]: Why not `SpendingCounterIncreasing::default`?

File: chain-impl-mockchain/src/accounting/account/spending.rs (hunk 1 above)
Comment (Zeegomo) [reply chain-impl-mockchain/src/accounting/account/spending.rs]: I meant that if n is large (e.g. `u32::MAX`), adding its value to the current counter in `spending_counter.increment_nth(n)` will panic in debug mode.

Comment (dkijania) [reply chain-impl-mockchain/src/accounting/account/spending.rs]: I Split test cases then to discard illegal case in quickcheck but added explicit case for overflows

**Review by Zeegomo — APPROVED** [review]

**Pull request merged**